### ISSUE TITLE
simplify does not simplify "Abs(cos(x))*sign(cos(x))" to "cos(x)"

### ISSUE THREAD

**Issue by bernamdc** [sympy/sympy#20402]
simplify(Abs(x)*sign(x))  # Returns: "x". Successful!
simplify(Abs(cos(x))*sign(cos(x)))  # Does not return: "cos(x)". Unsuccessful!

**Comment by theanshm** [comment]
This is because cosine function can either return a positive or negative value even if you set x as postive.

**Comment by oscarbenjamin** [comment]
Related to #20266

**Comment by Maelstrom6** [comment]
The following works for me:
```python
from sympy import *
x = symbols('x', real=True)
print(simplify(Abs(cos(x))*sign(cos(x))))
```
If no assumptions are given, SymPy assumes `x` is complex. I can only assume that when `x` is complex, `sign` becomes the identity operator and we have `|cos(x)| cos(x)`.

**Comment by bernamdc** [comment]
> The following works for me:
> 
> ```python
> from sympy import *
> x = symbols('x', real=True)
> print(simplify(Abs(cos(x))*sign(cos(x))))
> ```
> 
> If no assumptions are given, SymPy assumes `x` is complex. I can only assume that when `x` is complex, `sign` becomes the identity operator and we have `|cos(x)| cos(x)`.

For me it doesn't work, even when defining the variable as real (note: `from sympy import *` is deprecated)
I also believe this should also apply to complex numbers. Reference: [https://www.wolframalpha.com/input/?i=simplify%28abs%28a%2Bi*b%29*sgn%28a%2Bi*b%29%29](url)

**Comment by Maelstrom6** [comment]
I see version 1.6.1 fails to simplify but on the current master branch it does simplify. I assume 1.7 will be out somewhat soon. 

>note: from sympy import * is deprecated

I haven't seen this update. It's discouraged in the source code, but for user purposes, I'm sure it's just fine.

> I also believe this should also apply to complex numbers.

Ah, that's the first time I've learnt about `sign` with a complex argument. My apologies, I'm not sure why it doesn't work for complex numbers.

**Comment by meganly** [comment]
What specific simplification function does ``simplify`` use to simplify ``(Abs(x)*sign(x)``?

**Comment by Maelstrom6** [comment]
> What specific simplification function does simplify use to simplify Abs(x)*sign(x)?

It's difficult to pin down but this seems to have the biggest impact:
https://github.com/sympy/sympy/blob/6cd66656e6ce165609cb6245b7ef010b6d2777d7/sympy/simplify/simplify.py#L639-L644

Is this issue fixed? I tried the following on master and got `cos(x)`:
```python
from sympy import *
x = symbols('x')
print(simplify(Abs(cos(x))*sign(cos(x))))
```

**Comment by oscarbenjamin** [comment]
See also the discussion in #20552 where this leads to the less desirable:
```python
In [117]: simplify(x*sign(x))
Out[117]: 
⎧ 0   for x = 0
⎪              
⎪  2           
⎨ x            
⎪───  otherwise
⎪│x│           
⎩    
```
I think that it can be improved somehow.

**Comment by meganly** [comment]
> Is this issue fixed? I tried the following on master and got `cos(x)`:
> 
> ```python
> from sympy import *
> x = symbols('x')
> print(simplify(Abs(cos(x))*sign(cos(x))))
> ```

 Looks like it.  The sign function gets rewritten as Abs https://github.com/sympy/sympy/blob/6cd66656e6ce165609cb6245b7ef010b6d2777d7/sympy/simplify/simplify.py#L640-L641
the expression gets folded into a single Piecewise https://github.com/sympy/sympy/blob/6cd66656e6ce165609cb6245b7ef010b6d2777d7/sympy/simplify/simplify.py#L646
and the piecewise is eliminated https://github.com/sympy/sympy/blob/6cd66656e6ce165609cb6245b7ef010b6d2777d7/sympy/simplify/simplify.py#L662

This gives me ``cos(x)`` on master.

**Comment by oscarbenjamin** [comment]
A test should be added if this is working now.

I do think that the implementation of this can be improved though so that `simplify(x*sign(x))` doesn't give `Piecewise`.

**Comment by meganly** [comment]
> I do think that the implementation of this can be improved though so that `simplify(x*sign(x))` doesn't give `Piecewise`.

Definitely.  When `x` is real you can rewrite ``Abs`` as  ``sign(x)*x`` so you don't have to deal with a piecewise function.  I'll think more about what to do when `x` is complex.

**Comment by oscarbenjamin** [comment]
The definition sympy uses for `sign(x)` is basically just `x/Abs(x)` for any complex `x` (except 0):
```
In [14]: sign(1 + I).n()
Out[14]: 0.707106781186548 + 0.707106781186548⋅ⅈ
```

**Comment by meganly** [comment]
It seems like simplify is using
https://github.com/sympy/sympy/blob/6cd66656e6ce165609cb6245b7ef010b6d2777d7/sympy/functions/elementary/complexes.py#L438-L439
to rewrite the sign as an absolute value.  However the piecewise definition is causing undesirable results because most people wouldn't consider a piecewise function to be simpler.  Could this method just return `sign` as `x/Abs(x)`?  A lot of work is done later in the code to get rid of the piecewise definition and ignore the case when `x=0`.

**Comment by Maelstrom6** [comment]
> See also the discussion in #20552 where this leads to the less desirable:
> 
> ```python
> In [117]: simplify(x*sign(x))
> Out[117]: 
> ⎧ 0   for x = 0
> ⎪              
> ⎪  2           
> ⎨ x            
> ⎪───  otherwise
> ⎪│x│           
> ⎩    
> ```
> 
> I think that it can be improved somehow.

This example probably shows another problem. According to the `simplify` docs, we have:

> In some cases, applying :func:`simplify` may actually result in some more complicated expression. The default ``ratio=1.7`` prevents more extreme cases: if (result length)/(input length) > ratio, then input is returned unmodified.

But that does not happen with your example.

```python
>>> count_ops(x*sign(x))
2
>>> count_ops(simplify(x*sign(x)))
4
```

**Comment by meganly** [comment]
Pattern matching with ``Wild`` seems like a promising idea, but I do not understand why we need to preserve the behavior at zero.  There are many instances where simplifying or rewriting an expression does not preserve the behavior at a singularity.  For example:

```python
>>> sin(x).subs(x,pi)
0
>>> (sin(x).rewrite(sin, tan)).subs(x,pi)
nan
>>> ((x**2+5*x+6)/(x+3)).subs(x,-3)
nan
>>> simplify((x**2+5*x+6)/(x+3)).subs(x,-3)
-1
```

**Comment by meganly** [comment]
Would it be possible to change ``_eval_rewrite_as_Abs`` to return ``x/Abs(x)``?

The inconsistency of when Sympy cares about assumptions and when it doesn't makes it difficult to understand how rewrite methods behave.  I'm confused why some rewrite methods need to preserve the behavior at singularities and others don't.

This makes me concerned that ``rewrite`` might not rewrite a function as expected because it's making an assumption I'm not aware of or because I don't pass in the correct assumptions.

Is it be better to use Wild to do pattern matching and replacement for the substitutions I want instead of using rewrite?

For example, if I want to rewrite ``sign(x-1)`` as ``(x-1)/Abs(x-1)`` because I'm assuming x is not 1, then should I do ``sign(x-1).replace(sign(Wild('*')), Wild('*')/Abs(Wild('*')))`` instead of ``sign(x-1).rewrite(sign, Abs)``?

**Comment by oscarbenjamin** [comment]
The general practice (not always applied consistently) is that ideally `rewrite` would always return something that is equivalent wrt any assumptions that are given and handles any possible singularities by using `Piecewise`.

Cancelling denominators in simplify is a tricky case but I think it should at least always be a removable singularity in that case.

**Comment by oscarbenjamin** [comment]
> Is it better to use Wild to do pattern matching and replacement for the substitutions I want instead of using `rewrite`?

The most direct way to tell sympy *exactly* how you want to rewrite an expression is using `replace`.

**Comment by oscarbenjamin** [comment]
> > Is it better to use Wild to do pattern matching and replacement for the substitutions I want instead of using `rewrite`?
> 
> The most direct way to tell sympy _exactly_ how you want to rewrite an expression is using `replace`.

And then I read the next sentence where you did exactly that!

> For example, if I want to rewrite `sign(x-1)` as `(x-1)/Abs(x-1)` because I'm assuming x is not 1, then should I do `sign(x-1).replace(sign(Wild('*')), Wild('*')/Abs(Wild('*')))` instead of `sign(x-1).rewrite(sign, Abs)`?

The `replace` method is there when you just want to tell sympy exactly what to do. It is an instruction like "do exactly what I say and I don't care whether it is mathematically or structurally valid". It's always there as an option but it's up to you to use it in the way that you want.
